### DISCUSSION TITLE
Where do these comments come from in html code? <!-- -->

### DISCUSSION THREAD

**Discussion by ermakov95** [vercel/next.js#42612] [Help] [answered]
Example from my project: 
`<span class="styles_reviews__title__etUVQ"
                      >5
                      ·
                      2
                      reviews</span
                    >`
Why are these comments needed? And is there any way to remove them?

**Comment by icyJoseph** [comment, marked as answer]
This is how interpolations are generated as part of the React Server Side rendering process. Basically the thing that generates the initial HTML string. More at the bottom of this post.

Refer to my answer here, https://github.com/vercel/next.js/discussions/38256#discussioncomment-3070196

In short:

```jsx
const Home: NextPage = () => {
  const world = "World";
  return (
    <div>
      <span>Hello {world} foo</span>
    </div>
  );
};
```

Renders:

```html
<div><span>Hello <!-- -->World<!-- --> foo</span></div>
```

If it bothers you can generate the interpolated string in another way. For example:

```jsx
  const message = `Hello ${world} foo`;
  return (
    <div>
      <span>{message}</span>
    </div>
  );
```

or

```jsx
  <span>{`Hello ${world} foo`}</span>
```

## Who does this?

For example, you can also do this in a Create-React-App CodeSandBox, so Next.js is out of the way.

```js
import { renderToString } from 'react-dom/server'

import App from "./App";

console.log(renderToString(<App/>))
```

Where `App` is:

```js
import "./styles.css";

export default function App() {
  const foo = "foo";
  return (
    <div className="App">
      <h1>Hello CodeSandbox</h1>
      <h2>Start editing to see some magic happen! {foo}</h2>
    </div>
  );
}
```

Then in the console you'll see:

```html
<div class="App"><h1>Hello CodeSandbox</h1><h2>Start editing to see some magic happen! <!-- -->foo</h2></div> 
```

And there you see it! This is done, AFAIK, for hydration purposes, to see where the DOM might have diverged. Do remember that there's such a thing as Nodes, which include text and comments.
